compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/ImmutableTypeCloner.java

```diff
@@ -350,46 +352,9 @@ private static BIntersectionType setImmutableType(Location pos, Types types,
-                BAnydataType origAnydataType = (BAnydataType) type;
-
-                BTypeSymbol immutableAnydataTSymbol = getReadonlyTSymbol(names, origAnydataType.tsymbol, env, pkgId,
-                                                                         owner);
-
-                BAnydataType immutableAnydataType;
-                if (immutableAnydataTSymbol != null) {
-                    immutableAnydataType =
-                            new BAnydataType(immutableAnydataTSymbol,
-                                             immutableAnydataTSymbol.name, origAnydataType.flags | Flags.READONLY,
-                                             origAnydataType.isNullable());
-                    immutableAnydataTSymbol.type = immutableAnydataType;
-                } else {
-                    immutableAnydataType =
-                            new BAnydataType(immutableAnydataTSymbol,
-                                             getImmutableTypeName(names, TypeKind.ANYDATA.typeName()),
-                                             origAnydataType.flags | Flags.READONLY, origAnydataType.isNullable());
-                }
-
-                BIntersectionType immutableAnydataIntersectionType =
-                        createImmutableIntersectionType(pkgId, owner, originalType, immutableAnydataType, symTable);
-                Types.addImmutableType(symTable, pkgId, origAnydataType, immutableAnydataIntersectionType);
-                return immutableAnydataIntersectionType;
             case TypeTags.JSON:
-                BJSONType origJsonType = (BJSONType) type;
-
-                BTypeSymbol immutableJsonTSymbol = getReadonlyTSymbol(names, origJsonType.tsymbol, env, pkgId, owner);
-                BJSONType immutableJsonType = new BJSONType(immutableJsonTSymbol,
-                                                            origJsonType.isNullable(),
-                                                            origJsonType.flags | Flags.READONLY);
-                if (immutableJsonTSymbol != null) {
-                    immutableJsonTSymbol.type = immutableJsonType;
-                }
-
-                BIntersectionType immutableJsonIntersectionType = createImmutableIntersectionType(pkgId, owner,
-                                                                                                  originalType,
-                                                                                                  immutableJsonType,
-                                                                                                  symTable);
-                Types.addImmutableType(symTable, pkgId, origJsonType, immutableJsonIntersectionType);
-                return immutableJsonIntersectionType;
+                return defineImmutableBuiltInUnionType(pos, types, env, pkgId, owner, symTable, anonymousModelHelper,
+                                                       names, unresolvedTypes, (BUnionType) type, originalType);
```

Comment:
Can't we generate the immutable type for built-in union types and store it in the 'immutableTypeMaps' at the beginning? (before the compilation start)

Reply:
Yes, It's correct. But it helps maintain the logic clean and eliminates the need for checks to see if the immutable type for built-in types has been defined. 
(I think the immutable type of a built-in type is also a built-in type)

Reply:
Not necessarily. The intersection type is created in the module the user defines the intersection. Not in the root module. We'll still have to check if it is defined. 

Also `json` is a built-in type, not `json & readonly`.